app-creator/src/redux/actions.ts

```diff
@@ -17,8 +18,77 @@ import {
   SET_REORDERING,
   IncrementWidgetAction,
   INCREMENT_WIDGET_ID,
+  AddWidgetMetaData,
+  RemoveWidgetMetaData,
+  UPDATE_WIDGET_META_DATA,
+  UpdateWidgetMetaData,
 } from './types/actions';
-import { EventType } from './reducer';
+import {
+  DEFAULT_SHARED_ATTRIBUTES,
+  UniqueAttributes,
+  DEFAULT_LABEL_ATTRIBUTES,
+  DEFAULT_BUTTON_ATTRIBUTES,
+  DEFAULT_SELECT_ATTRIBUTES,
+  DEFAULT_CHECKBOX_ATTRIBUTES,
+  DEFAULT_SLIDER_ATTRIBUTES,
+  DEFAULT_TEXTBOX_ATTRIBUTES,
+} from './types/attributes';
+import { WidgetType, AttributeType, Tab, EventType } from './types/enums';
+
+/**
+ * Updates widget attributes.
+ */
+export const updateWidgetMetaData = (
+  key: string,
+  value: string,
+  id: string,
+  attributeType: AttributeType
+): UpdateWidgetMetaData => {
+  return {
+    type: UPDATE_WIDGET_META_DATA,
+    payload: {
+      key,
+      value,
+      id,
+      attributeType,
+    },
+  };
+};
+
+/**
+ * Removes the widget metadata for the given widget id.
+ */
+export const removeWidgetMetaData = (id: string): RemoveWidgetMetaData => {
+  return {
+    type: REMOVE_WIDGET_META_DATA,
+    payload: {
+      id,
+    },
+  };
+};
+
+/**
+ * Adds widget meta data to the store's template representation.
+ */
+export const addWidgetMetaData = (
+  id: string,
+  widget: Element
+): AddWidgetMetaData => {
+  return {
+    type: ADD_WIDGET_META_DATA,
+    payload: {
+      [id]: {
+        id,
+        widgetRef: widget as HTMLElement,
+        children: [],
+        uniqueAttributes: {
+          ...getDefaultUniqueAttributes(id.slice(0, id.indexOf('-'))),
```

Comment:
Can we replace `id.slice(0, id.indexOf('-'))` with a get ID method, just in case it needs to be reused, or the way we format widget ID is changed?

Reply:
Perhaps you can use the widget element's base class name?